File: dt.app-laravel/dt.app-main/app/Models/EdiRecord.php
<?php

namespace App\Models;

class EdiRecord
{
    const FIELDS = [
        'bl_number'                     => [62,  20],
        'import_export'                 => [null, 0],
        'import_export_raw'             => [694, 10],
        'stevedore'                     => [null, 0],
        'shipping_agent'                => [null, 0],
        'estimated_departure_date'      => [null, 0],
        'call_number'                   => [21,  10],
        'shipper'                       => [null, 0],
        'forwarder'                     => [null, 0],
        'related_customer'              => [null, 0],
        'forwarding_agent'              => [null, 0],
        'final_destination_country'     => [null, 0],
        'manifest'                      => [926, 35],
        'number_of_yard_items'          => [1757, 5],
        'number_of_packages'            => [1757, 5],
        'slot_file'                     => [null, 0],
        'transport_mode'                => [61,   1],
        'consignee'                     => [926, 35],
        'custom_release_order'          => [null, 0],
        'custom_release_order_date'     => [null, 0],
        'delivery_order'                => [null, 0],
        'delivery_order_date'           => [null, 0],
        'master_bl'                     => [null, 0],
        'bl_volume'                     => [293, 12],   // conteneurs
        'bl_volume_roro'                => [1308, 12],  // véhicules RORO
        'bl_weight'                     => [281, 12],
        'bl_weight_alt'                 => [1296, 12], // poids alternatif (certains RORO / conteneurs)
        'incoterm'                      => [null, 0],
        'port_of_loading'               => [31,   5],
        'reception_location'            => [46,   5],
        'transshipment_port_1'          => [36,   5],
        'transshipment_port_2'          => [41,   5],
        'commodity'                     => [null, 0],
        'yard_item_type'                => [null, 0],
        'unit_of_measure'               => [null, 0],
        'comment'                       => [null, 0],
        'direction_code'                => [null, 0],
        'agent_name'                    => [1111, 35],
        'blitem_yard_item_type'         => [null, 0],
        'blitem_comment'                => [221, 35],
        'blitem_yard_item_number'       => [126, 20],
        'blitem_allow_invalid'          => [null, 0],
        'blitem_yard_item_code'         => [146, 35],
        'blitem_out_of_gauge'           => [null, 0],
        'blitem_commodity'              => [null, 0],
        'blitem_unloading_date'         => [null, 0],
        'blitem_commodity_volume'       => [null, 0],
        'blitem_commodity_weight'       => [281, 12],
        'blitem_commodity_packages'     => [null, 0],
        'blitem_import_export'          => [null, 0],
        'blitem_custom_number'          => [null, 0],
        'blitem_seal_number_1'          => [351, 70],   // 70 chars pour capturer les deux sceaux séparés par |
        'blitem_seal_number_2'          => [null, 0],   // déduit du champ 1
        'blitem_commodity_hazardous_class' => [null, 0],
        'blitem_barcode'                => [126, 20],
        'blitem_vehicle_model'          => [221, 35],
        'blitem_chassis_number'         => [126, 20],
        'outgoing_call_number'          => [null, 0],
        'outgoing_slot_file'            => [null, 0],
        'is_lifter'                     => [null, 0],
        'stacked_chassis'               => [null, 0],
        'stacked_model'                 => [null, 0],
        'stacked_weight'                => [null, 0],
        'stacked_volume'                => [null, 0],
        'new_transshipment_bl'          => [null, 0],
        'shipper_name'                  => [741, 35],
        'adresse_2'                     => [961, 35],
        'adresse_3'                     => [996, 35],
        'adresse_4'                     => [1031, 35],
        'adresse_5'                     => [1066, 35],
        'notify1'                       => [1111, 35],
        'notify2'                       => [1146, 35],
        'notify3'                       => [1181, 35],
        'notify4'                       => [1216, 35],
        'notify5'                       => [1251, 35],
    ];

    public array $data = [];

    public static function fromLine(string $line, string $srcEncoding = 'UTF-8'): self
    {
        $record = new self();

        foreach (self::FIELDS as $field => [$offset, $length]) {
            if ($offset === null || $length === 0) {
                $record->data[$field] = '';
                continue;
            }
            // substr() works on raw bytes — correct for fixed-width single-byte encodings.
            $value = $offset < strlen($line) ? substr($line, $offset, $length) : '';
            // Convert to UTF-8 per field so byte offsets stay accurate for all fields.
            if ($srcEncoding !== 'UTF-8' && $value !== '') {
                $value = mb_convert_encoding($value, 'UTF-8', $srcEncoding);
            }
            $record->data[$field] = trim($value);
        }

        // ── BL Number : conserver tel quel ───────────────────────────────────
        $record->data['bl_number'] = trim($record->data['bl_number']);

        // ── Poids : offset 281 (principal), fallback offset 1296 ─────────────
        $rawWeight281 = ltrim($record->data['bl_weight']     ?? '', '0') ?: '0';
        $rawWeight296 = ltrim($record->data['bl_weight_alt'] ?? '', '0') ?: '0';
        if (is_numeric($rawWeight281) && (int)$rawWeight281 > 0) {
            $weight = round((float)$rawWeight281 / 1_000_000, 6);
        } elseif (is_numeric($rawWeight296) && (int)$rawWeight296 > 0) {
            $weight = round((float)$rawWeight296 / 1_000_000, 6);
        } else {
            $weight = 0;
        }
        $weightStr = $weight > 0 ? (string)$weight : '';
        $record->data['bl_weight']               = $weightStr;
        $record->data['blitem_commodity_weight'] = $weightStr;
        unset($record->data['bl_weight_alt']);

        // ── Volume : offset 293 (conteneurs) ou 1308 (véhicules RORO) ────────
        $rawVol293  = ltrim($record->data['bl_volume']      ?? '', '0') ?: '0';
        $rawVol1308 = ltrim($record->data['bl_volume_roro'] ?? '', '0') ?: '0';
        $vol = 0;
        if (is_numeric($rawVol293) && (int)$rawVol293 > 0) {
            $vol = round((int)$rawVol293 / 1000, 3);
        } elseif (is_numeric($rawVol1308) && (int)$rawVol1308 > 0) {
            $vol = round((int)$rawVol1308 / 1000, 3);
        }
        $volStr = $vol > 0 ? (string)$vol : '';
        $record->data['bl_volume']               = $volStr;
        $record->data['blitem_commodity_volume'] = $volStr;
        unset($record->data['bl_volume_roro']);

        // ── Nombre d'articles ─────────────────────────────────────────────────
        $nyi = trim($record->data['number_of_yard_items']);
        $record->data['number_of_yard_items'] = is_numeric($nyi) ? $nyi : '1';
        $record->data['number_of_packages']   = $record->data['number_of_yard_items'];

        // ── ImportExport ──────────────────────────────────────────────────────
        $raw = trim($record->data['import_export_raw'] ?? '');
        if (str_starts_with($raw, 'TS')) {
            $record->data['import_export'] = 'TRANSBO';
        } elseif (str_starts_with($raw, 'TR')) {
            $record->data['import_export'] = 'IMPORT TRANSIT';
        } else {
            $record->data['import_export'] = 'IMPORT';
        }
        unset($record->data['import_export_raw']);

        // ── Final Destination Country : reprend adresse_5 ────────────────────
        $record->data['final_destination_country'] = trim($record->data['adresse_5'] ?? '');

        // ── YardItemType : déduit du mode de transport ────────────────────────
        $mode     = trim($record->data['transport_mode'] ?? '');
        $yardType = in_array($mode, ['R', 'M']) ? 'VEHICULE' : 'CONTENEUR';
        $record->data['yard_item_type']        = $yardType;
        $record->data['blitem_yard_item_type'] = $yardType;

        // ── BLItem AllowInvalidYardItemNumber : toujours VRAI ─────────────────
        $record->data['blitem_allow_invalid'] = 'VRAI';

        // ── BLItem Commodity : tranches de poids (véhicules uniquement) ───────
        // $weight est en tonnes (÷1 000 000). Seuils convertis en tonnes.
        if ($yardType === 'VEHICULE' && $weight > 0) {
            $record->data['blitem_commodity'] = match (true) {
                $weight <= 1.5   => 'VEH 0-1500Kgs',
                $weight <= 3.0   => 'VEH 1501-3000Kgs',
                $weight <= 6.0   => 'VEH 3001-6000Kgs',
                $weight <= 9.0   => 'VEH 6001-9000Kgs',
                $weight <= 30.0  => 'VEH 9001-30000Kgs',
                default          => 'VEH +30000Kgs',
            };
        } else {
            $record->data['blitem_commodity'] = '';
        }

        // ── Sceaux : format différent selon le type de transport ─────────────
        // RORO (mode R/M) : le champ 70-chars contient deux sceaux séparés par |
        //   → on split : seal1 = premier sceau + '|', seal2 = second sceau + '|'
        // Conteneurs (autres modes) : le champ contient la valeur complète avec |
        //   → on garde tel quel dans seal1, seal2 = seal1 (BUG-E)
        $sealRaw = trim($record->data['blitem_seal_number_1'] ?? '');
        if (in_array($mode, ['R', 'M'])) {
            // RORO : split at |
            $sealParts = array_values(array_filter(array_map('trim', explode('|', $sealRaw))));
            $record->data['blitem_seal_number_1'] = isset($sealParts[0]) ? $sealParts[0] . '|' : '';
            $record->data['blitem_seal_number_2'] = isset($sealParts[1]) ? $sealParts[1] . '|' : '';
        } else {
            // Conteneurs : garder la valeur brute complète
            $record->data['blitem_seal_number_1'] = $sealRaw;
            $record->data['blitem_seal_number_2'] = $sealRaw;
        }

        // ── New template fields (always empty – no source offset available) ──
        $record->data['blitem_hs_code']           = '';
        $record->data['blitem_gross_weight']      = '';
        $record->data['freight_prepaid_collect']  = '';
        $record->data['shipping_line_export_bl']  = '';
        $record->data['is_transfer']              = '';
        $record->data['blitem_hazardous_class']   = '';
        $record->data['attach_to_bl']             = '';

        return $record;
    }

    public function toArray(): array
    {
        return $this->data;
    }
}

File: src/main/java/com/dtapp/service/EdiRecord.java
package com.dtapp.service;

import java.nio.charset.Charset;
import java.util.Locale;
import java.util.*;

/**
 * Représente un enregistrement EDI (une ligne du fichier TXT à largeur fixe).
 * Traduit exactement la logique PHP de App\Models\EdiRecord.
 */
public class EdiRecord {

    /** [offset, length] – null = champ dérivé (pas lu directement). */
    public static final Map<String, int[]> FIELDS;

    static {
        Map<String, int[]> f = new LinkedHashMap<>();
        f.put("bl_number",                       new int[]{62,   20});
        f.put("import_export",                   null);
        f.put("import_export_raw",               new int[]{694,  10});
        f.put("stevedore",                       null);
        f.put("shipping_agent",                  null);
        f.put("estimated_departure_date",        null);
        f.put("call_number",                     new int[]{21,   10});
        f.put("shipper",                         null);
        f.put("forwarder",                       null);
        f.put("related_customer",                null);
        f.put("forwarding_agent",                null);
        f.put("final_destination_country",       null);
        f.put("manifest",                        new int[]{926,  35});
        f.put("number_of_yard_items",            new int[]{1757,  5});
        f.put("number_of_packages",              new int[]{1757,  5});
        f.put("slot_file",                       null);
        f.put("transport_mode",                  new int[]{61,    1});
        f.put("consignee",                       new int[]{926,  35});
        f.put("custom_release_order",            null);
        f.put("custom_release_order_date",       null);
        f.put("delivery_order",                  null);
        f.put("delivery_order_date",             null);
        f.put("master_bl",                       null);
        f.put("bl_volume",                       new int[]{293,  12}); // conteneurs
        f.put("bl_volume_roro",                  new int[]{1308, 12}); // véhicules RORO
        f.put("bl_weight",                       new int[]{281,  12});
        f.put("bl_weight_alt",                   new int[]{1296, 12}); // poids alternatif
        f.put("incoterm",                        null);
        f.put("port_of_loading",                 new int[]{31,    5});
        f.put("reception_location",              new int[]{46,    5});
        f.put("transshipment_port_1",            new int[]{36,    5});
        f.put("transshipment_port_2",            new int[]{41,    5});
        f.put("commodity",                       null);
        f.put("yard_item_type",                  null);
        f.put("unit_of_measure",                 null);
        f.put("comment",                         null);
        f.put("direction_code",                  null);
        f.put("agent_name",                      new int[]{1111, 35});
        f.put("blitem_yard_item_type",           null);
        f.put("blitem_comment",                  new int[]{221,  35});
        f.put("blitem_yard_item_number",         new int[]{126,  20});
        f.put("blitem_allow_invalid",            null);
        f.put("blitem_yard_item_code",           new int[]{146,  35});
        f.put("blitem_out_of_gauge",             null);
        f.put("blitem_commodity",                null);
        f.put("blitem_unloading_date",           null);
        f.put("blitem_commodity_volume",         null);
        f.put("blitem_commodity_weight",         new int[]{281,  12});
        f.put("blitem_commodity_packages",       null);
        f.put("blitem_import_export",            null);
        f.put("blitem_custom_number",            null);
        f.put("blitem_seal_number_1",            new int[]{351,  70}); // 70 chars pour deux sceaux séparés par |
        f.put("blitem_seal_number_2",            null);                 // déduit du champ 1
        f.put("blitem_commodity_hazardous_class", null);
        f.put("blitem_barcode",                  new int[]{126,  20});
        f.put("blitem_vehicle_model",            new int[]{221,  35});
        f.put("blitem_chassis_number",           new int[]{126,  20});
        f.put("outgoing_call_number",            null);
        f.put("outgoing_slot_file",              null);
        f.put("is_lifter",                       null);
        f.put("stacked_chassis",                 null);
        f.put("stacked_model",                   null);
        f.put("stacked_weight",                  null);
        f.put("stacked_volume",                  null);
        f.put("new_transshipment_bl",            null);
        f.put("shipper_name",                    new int[]{741,  35});
        f.put("adresse_2",                       new int[]{961,  35});
        f.put("adresse_3",                       new int[]{996,  35});
        f.put("adresse_4",                       new int[]{1031, 35});
        f.put("adresse_5",                       new int[]{1066, 35});
        f.put("notify1",                         new int[]{1111, 35});
        f.put("notify2",                         new int[]{1146, 35});
        f.put("notify3",                         new int[]{1181, 35});
        f.put("notify4",                         new int[]{1216, 35});
        f.put("notify5",                         new int[]{1251, 35});
        FIELDS = Collections.unmodifiableMap(f);
    }

    /** Données extraites et dérivées de la ligne. */
    public final Map<String, String> data = new LinkedHashMap<>();

    /**
     * Construit un EdiRecord à partir des bytes bruts d'une ligne.
     * Les offsets s'appliquent sur les bytes bruts (encodage à octet unique).
     */
    public static EdiRecord fromLine(byte[] raw, Charset srcCharset) {
        EdiRecord rec = new EdiRecord();

        // ── Extraction brute par offset ───────────────────────────────────────
        for (Map.Entry<String, int[]> e : FIELDS.entrySet()) {
            String field = e.getKey();
            int[] def    = e.getValue();
            if (def == null) {
                rec.data.put(field, "");
                continue;
            }
            int offset = def[0];
            int length = def[1];
            String value = "";
            if (offset < raw.length) {
                int end = Math.min(offset + length, raw.length);
                value = new String(raw, offset, end - offset, srcCharset).trim();
            }
            rec.data.put(field, value);
        }

        // ── BL Number ────────────────────────────────────────────────────────
        rec.data.put("bl_number", rec.data.getOrDefault("bl_number", "").trim());

        // ── Poids : offset 281 (principal), fallback 1296 — valeur interne en tonnes ────────
        String rawW281 = stripLeadingZeros(rec.data.getOrDefault("bl_weight", ""));
        String rawW296 = stripLeadingZeros(rec.data.getOrDefault("bl_weight_alt", ""));
        double weight  = 0;
        if (isPositiveNumeric(rawW281)) {
            weight = roundTo(Double.parseDouble(rawW281) / 1_000_000.0, 6);
        } else if (isPositiveNumeric(rawW296)) {
            weight = roundTo(Double.parseDouble(rawW296) / 1_000_000.0, 6);
        }
        String weightStr = weight > 0 ? formatDouble(weight) : "";
        rec.data.put("bl_weight",               weightStr);
        rec.data.put("blitem_commodity_weight", weightStr);
        rec.data.remove("bl_weight_alt");

        // ── Volume : offset 293 (conteneurs) ou 1308 (RORO) ─────────────────
        String rawV293  = stripLeadingZeros(rec.data.getOrDefault("bl_volume", ""));
        String rawV1308 = stripLeadingZeros(rec.data.getOrDefault("bl_volume_roro", ""));
        double vol = 0;
        if (isPositiveNumeric(rawV293)) {
            vol = roundTo(Double.parseDouble(rawV293) / 1000.0, 3);
        } else if (isPositiveNumeric(rawV1308)) {
            vol = roundTo(Double.parseDouble(rawV1308) / 1000.0, 3);
        }
        String volStr = vol > 0 ? formatDouble(vol) : "";
        rec.data.put("bl_volume",               volStr);
        rec.data.put("blitem_commodity_volume", volStr);
        rec.data.remove("bl_volume_roro");

        // ── Nombre d'articles ─────────────────────────────────────────────────
        String nyi = rec.data.getOrDefault("number_of_yard_items", "").trim();
        String nyiVal = isNumeric(nyi) ? nyi : "1";
        rec.data.put("number_of_yard_items", nyiVal);
        rec.data.put("number_of_packages",   nyiVal);

        // ── ImportExport ──────────────────────────────────────────────────────
        String rawIE = rec.data.getOrDefault("import_export_raw", "").trim();
        String importExport = rawIE.startsWith("TS") ? "TRANSBO"
                            : rawIE.startsWith("TR") ? "IMPORT TRANSIT"
                            : "IMPORT";
        rec.data.put("import_export", importExport);
        rec.data.remove("import_export_raw");

        // ── Final Destination Country = adresse_5 ────────────────────────────
        rec.data.put("final_destination_country", rec.data.getOrDefault("adresse_5", "").trim());

        // ── YardItemType ──────────────────────────────────────────────────────
        String mode     = rec.data.getOrDefault("transport_mode", "").trim();
        String yardType = (mode.equals("R") || mode.equals("M")) ? "VEHICULE" : "CONTENEUR";
        rec.data.put("yard_item_type",       yardType);
        rec.data.put("blitem_yard_item_type", yardType);

        // ── AllowInvalid ──────────────────────────────────────────────────────
        rec.data.put("blitem_allow_invalid", "VRAI");

        // ── Commodity (tranches de poids véhicules, weight en tonnes) ────────
        if ("VEHICULE".equals(yardType) && weight > 0) {
            String commodity;
            if      (weight <= 1.5)   commodity = "VEH 0-1500Kgs";
            else if (weight <= 3.0)   commodity = "VEH 1501-3000Kgs";
            else if (weight <= 6.0)   commodity = "VEH 3001-6000Kgs";
            else if (weight <= 9.0)   commodity = "VEH 6001-9000Kgs";
            else if (weight <= 30.0)  commodity = "VEH 9001-30000Kgs";
            else                      commodity = "VEH +30000Kgs";
            rec.data.put("blitem_commodity", commodity);
        } else {
            rec.data.put("blitem_commodity", "");
        }

        // ── Sceaux ────────────────────────────────────────────────────────────
        String sealRaw = rec.data.getOrDefault("blitem_seal_number_1", "").trim();
        if (mode.equals("R") || mode.equals("M")) {
            // RORO : séparer sur |
            String[] parts = Arrays.stream(sealRaw.split("\\|"))
                    .map(String::trim).filter(s -> !s.isEmpty()).toArray(String[]::new);
            rec.data.put("blitem_seal_number_1", parts.length > 0 ? parts[0] + "|" : "");
            rec.data.put("blitem_seal_number_2", parts.length > 1 ? parts[1] + "|" : "");
        } else {
            rec.data.put("blitem_seal_number_1", sealRaw);
            rec.data.put("blitem_seal_number_2", sealRaw);
        }

        // ── Champs vides (pas de source dans le TXT) ─────────────────────────
        rec.data.put("blitem_hs_code",          "");
        rec.data.put("blitem_gross_weight",     "");
        rec.data.put("freight_prepaid_collect", "");
        rec.data.put("shipping_line_export_bl", "");
        rec.data.put("is_transfer",             "");
        rec.data.put("blitem_hazardous_class",  "");
        rec.data.put("attach_to_bl",            "");

        return rec;
    }

    public Map<String, String> toArray() { return data; }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private static String stripLeadingZeros(String s) {
        if (s == null || s.isEmpty()) return "0";
        String r = s.replaceAll("^0+", "");
        return r.isEmpty() ? "0" : r;
    }

    static boolean isPositiveNumeric(String s) {
        if (s == null || s.isEmpty() || s.equals("0")) return false;
        try { return Double.parseDouble(s) > 0; } catch (NumberFormatException e) { return false; }
    }

    static boolean isNumeric(String s) {
        if (s == null || s.isEmpty()) return false;
        try { Double.parseDouble(s); return true; } catch (NumberFormatException e) { return false; }
    }

    static double roundTo(double value, int decimals) {
        double factor = Math.pow(10, decimals);
        return Math.round(value * factor) / factor;
    }

    static String formatDouble(double d) {
        // Remove trailing zeros like PHP does
        String s = String.format(Locale.ROOT, "%.6f", d);
        s = s.replaceAll("0+$", "").replaceAll("\\.$", "");
        return s;
    }
}
